### PR TITLE
Implement wild card contracts selector

### PR DESCRIPTION
**Pull request by ametel01**
solve #577 

only one test added, do we need to test more complex path declarations?

### CODE REVIEW

File: scarb/src/compiler/compilers/starknet_contract.rs [hunk 1]
```diff
@@ -297,32 +306,40 @@ fn find_project_contracts(
...
+                .collect::<Vec<_>>();
+            let contracts = find_contracts(db, crate_ids.as_ref());
+            let filtered_contracts: Vec<ContractDeclaration> = contracts
+                .into_iter()
+                .filter(|decl| {
                     let contract_path = decl.module_id().full_path(db.upcast());
-                    contract_path == selector.full_path()
+                    external_contracts.iter().any(|selector| {
+                        if selector.is_wildcard() {
+                            contract_path.starts_with(&selector.package_path())
```
Comment (maciektr) [inline scarb/src/compiler/compilers/starknet_contract.rs:330]: I do not think this logic is correct. 
As of now, selector with `some::cairo::path::*` would match all contracts in `some` package.

File: scarb/src/compiler/compilers/starknet_contract.rs [hunk 2]
```diff
@@ -78,6 +78,15 @@ impl ContractSelector {
         PackageName::new(parts.0)
     }
 
+    fn is_wildcard(&self) -> bool {
+        self.0.ends_with("::*")
```
Comment (maciektr) [inline scarb/src/compiler/compilers/starknet_contract.rs:82]: Please refer to `::` as `CAIRO_PATH_SEPARATOR` wherever possible.

File: scarb/src/compiler/compilers/starknet_contract.rs [hunk 3]
```diff
@@ -78,6 +78,15 @@ impl ContractSelector {
...
+    }
+
+    fn package_path(&self) -> String {
+        let parts = self.0.split_once("::").unwrap_or((self.0.as_str(), ""));
```
Comment (maciektr) [inline scarb/src/compiler/compilers/starknet_contract.rs:86]: It's equivalent to `.package`.

File: scarb/src/compiler/compilers/starknet_contract.rs (hunk 1 above)
Comment (maciektr) [reply scarb/src/compiler/compilers/starknet_contract.rs:330]: Please also add a test case for this scenario.

File: scarb/src/compiler/compilers/starknet_contract.rs (hunk 2 above)
Comment (ametel01) [reply scarb/src/compiler/compilers/starknet_contract.rs:82]: implemented GLOB_PATH_SELECTOR const

File: scarb/src/compiler/compilers/starknet_contract.rs (hunk 1 above)
Comment (ametel01) [reply scarb/src/compiler/compilers/starknet_contract.rs:330]: change package_path to partial_path selector with modified logic

File: scarb/src/compiler/compilers/starknet_contract.rs (hunk 2 above)
Comment (ametel01) [reply scarb/src/compiler/compilers/starknet_contract.rs:82]: added GLOB_PATH_SEPARATOR const

File: scarb/src/compiler/compilers/starknet_contract.rs (hunk 3 above)
Comment (ametel01) [reply scarb/src/compiler/compilers/starknet_contract.rs:86]: removed

**Comment by mkaput** [comment]
> cargo fmt --check passing locally not sure why not on CI

Rust 1.72 got released and it changed formatting a bit

**Comment by ametel01** [comment]
> > cargo fmt --check passing locally not sure why not on CI
> 
> Rust 1.72 got released and it changed formatting a bit
Updated rust and now all good

**Review by maciektr — COMMENTED** [review]
Great work! 🔥 🔥 🔥 

One last thing to consider: 
Can we also verify that the path selector contains no more than a single `*`?
Consider emitting an Err result with https://docs.rs/anyhow/latest/anyhow/macro.ensure.html otherwise.

**Comment by ametel01** [comment]
> Great work! 🔥 🔥 🔥
> 
> One last thing to consider: Can we also verify that the path selector contains no more than a single `*`? Consider emitting an Err result with https://docs.rs/anyhow/latest/anyhow/macro.ensure.html otherwise.

What kind of behaviour do you want in case ensure! doesn't pass? return anyhow::Error or log a message and continue with the function?

**Comment by maciektr** [comment]
@ametel01 Since we implement prefixes only, I think we should ensure that the package selector has exactly one `*` or none and that it is the last character, as otherwise behaviour of `some::path::*::with::two::*` may be confusing to users. This should be an error.

**Comment by ametel01** [comment]
I’m unsure where to put the check for ‘*’ selector is the last character of the path since since the is_wildcard() only returns true when ‘*’ is last.

File: scarb/src/compiler/compilers/starknet_contract.rs [hunk 4]
```diff
@@ -291,38 +305,53 @@ fn find_project_contracts(
...
+                            }
+                        } else {
+                            contract_path == selector.full_path()
+                        }
+                    });
+                    if is_matched {
+                        Some(Ok::<ContractDeclaration, anyhow::Error>(decl))
+                    } else {
+                        None
+                    }
```
Comment (maciektr) [inline scarb/src/compiler/compilers/starknet_contract.rs:347]: I would rather show the parsing error to the user, so he can adjust his config appropriately. 

My suggestion would be, to validate the struct `Prop` which deserializes this value. 
https://github.com/software-mansion/scarb/blob/0db29e2b584b624a3f0fc4b8026d2aae3dfc0113/scarb/src/compiler/compilers/starknet_contract.rs#L186C9-L186C9
This way we could check all the selectors once,  and show the errors to the user. 

We can also have a simple test case with invalid selector to verify the validation.

File: scarb/src/compiler/compilers/starknet_contract.rs [hunk 5]
```diff
@@ -86,6 +87,19 @@ impl ContractSelector {
         parts.1.to_string()
     }
 
+    fn is_wildcard(&self) -> bool {
+        self.0.ends_with(GLOB_PATH_SELECTOR)
+    }
+
+    fn partial_path(&self) -> Result<String, anyhow::Error> {
+        ensure!(self.0.matches(GLOB_PATH_SELECTOR).count() <= 1, format!("External dependency {} has multiple global path selectors; only one '*' selector is allowed.", self.0));
```
Comment (maciektr) [inline scarb/src/compiler/compilers/starknet_contract.rs:95]: Nit
```suggestion
        ensure!(self.0.matches(GLOB_PATH_SELECTOR).count() <= 1, format!("external contract path {} has multiple global path selectors, only one '*' selector is allowed", self.0));
```

**Review by maciektr — COMMENTED** [review]
Hi!

Left one more small comment. 
 
Sorry for responding so late to your changes, I has been ooo for most of the week. 
If you have moved on to another challange already, please let me know.

File: scarb/src/compiler/compilers/starknet_contract.rs (hunk 4 above)
Comment (ametel01) [reply scarb/src/compiler/compilers/starknet_contract.rs]: removed all this, now ensure is enforced on every path of the external-contracts in `compile`

File: scarb/src/compiler/compilers/starknet_contract.rs (hunk 5 above)
Comment (ametel01) [reply scarb/src/compiler/compilers/starknet_contract.rs]: fixed

**Review by maciektr — APPROVED** [review]
This looks great! Thank you a lot for your help!

♥️♥️♥️♥️♥️

**Comment by maciektr** [comment]
Would you mind squashing the commits so I can merge this?

**Comment by ametel01** [comment]
> Would you mind squashing the commits so I can merge this?

Unable to squash commits, likely due to merging `main` into this branch.

**Comment by mkaput** [comment]
Hmm, I think interactive rebasing on top of main should purge the merge commit and give you an ability to squash these comments. Nonetheless, I think I can squash merge this manually if you all like :)

**Comment by ametel01** [comment]
> Hmm, I think interactive rebasing on top of main should purge the merge commit and give you an ability to squash these comments. Nonetheless, I think I can squash merge this manually if you all like :)

I attempted interactive rebase to remove the merge commit and squash the changes, but it complicated things further. Could you handle the squash merge on your end? Thanks!

**Comment by ametel01** [comment]
I think i managed somehow.

**Comment by mkaput** [comment]
OK, I will squash merge this anyway, so that we'll have a more descriptive commit message 😅 I'm sorry for mental shortcuts here; we generally use rebase merges & do as much as possible to maintain a linear and bisectable commit history 😄 But no worries, sorry for bothering.

I'm super thankful for this work and I'm happy to merge it!

❤️❤️❤️❤️❤️

**Pull request merged**